gemma/gemini_api_server.cc

```diff
@@ -0,0 +1,510 @@
+          
+          // Stream token callback
+          std::string accumulated_text;
+          auto stream_token = [&](int token, float) {
+            // Skip prompt tokens
+            if (session.abs_pos < tokens.size()) {
+              session.abs_pos++;
+              return true;
+            }
+            
+            session.abs_pos++;
+            
+            // Check for EOS
+            if (state.gemma->Config().IsEOS(token)) {
+              return true;
+            }
+            
+            // Decode token
+            std::string token_text;
+            state.gemma->Tokenizer().Decode(std::vector<int>{token}, &token_text);
+            accumulated_text += token_text;
+            
+            // Send SSE event using unified formatter
+            json event = CreateGeminiResponse(token_text, true);
+            
+            std::string sse_data = "data: " + event.dump() + "\n\n";
+            sink.write(sse_data.data(), sse_data.size());
+            
+            return true;
+          };
+          
+          runtime_config.stream_token = stream_token;
+          
+          // Run inference with KV cache
+          TimingInfo timing_info = {.verbosity = 0};
+          size_t prefix_end = 0;
+          
+          state.gemma->Generate(runtime_config, tokens, session.abs_pos, prefix_end, 
+                               *session.kv_cache, *state.env, timing_info);
+          
+          // Send final event using unified formatter
+          json final_event = CreateGeminiResponse("", false);
+          final_event["usageMetadata"] = {
+            {"promptTokenCount", tokens.size()},
+            {"candidatesTokenCount", session.abs_pos - tokens.size()},
+            {"totalTokenCount", session.abs_pos}
+          };
+          
+          std::string final_sse = "data: " + final_event.dump() + "\n\n";
+          sink.write(final_sse.data(), final_sse.size());
+          
+          // Send done event
+          sink.write("data: [DONE]\n\n", 15);
+          
+          // Ensure all data is sent
+          sink.done();
+          
+          return false; // End streaming
+          
+        } catch (const std::exception& e) {
+          json error_event = {{"error", {{"message", e.what()}}}};
+          std::string error_sse = "data: " + error_event.dump() + "\n\n";
+          sink.write(error_sse.data(), error_sse.size());
+          return false;
+        }
+      }
+    );
+    
+  } catch (const json::exception& e) {
+    res.status = 400;
+    res.set_content(json{{"error", {{"message", std::string("JSON parsing error: ") + e.what()}}}}.dump(), 
+                    "application/json");
+  }
+}
+
+// Handle models list endpoint
+void HandleListModels(ServerState& state, const httplib::Request& req, httplib::Response& res) {
+  json response = {
+    {"models", {{
+      {"name", "models/gemma-2b"},
+      {"version", "001"},
+      {"displayName", "Gemma 2B"},
```

Comment:
Is it an option to default to Gemma3 4B? It is likely much more capable.

Reply:
For now, the default becomes gemma3-4b, but can be changed as an option.